matrix_sdk_appservice/src/lib.rs

```diff
@@ -173,29 +175,13 @@ impl Deref for AppserviceRegistration {
     }
 }
 
-async fn client_session_with_login_restore(
-    client: &Client,
-    registration: &AppserviceRegistration,
-    localpart: impl AsRef<str> + Into<Box<str>>,
-    server_name: &ServerNameBox,
-) -> Result<()> {
-    let session = Session {
-        access_token: registration.as_token.clone(),
-        user_id: UserId::parse_with_server_name(localpart, server_name)?,
-        device_id: DeviceId::new(),
-    };
-    client.restore_login(session).await?;
-
-    Ok(())
-}
-
 /// Appservice
 #[derive(Debug, Clone)]
 pub struct Appservice {
     homeserver_url: Url,
     server_name: ServerNameBox,
-    registration: AppserviceRegistration,
-    client_sender_localpart: Client,
+    registration: Arc<AppserviceRegistration>,
+    clients: HashMap<UserId, Client>,
```

Comment:
Is this right? A clone will create a new `HashMap` here, and if you modify stuff between the clones they won't be synced anymore.

I think you'll want a `Arc<DashMap<UserId, Client>>` here.

Reply:
Good point, changed